File: 2024/25/25.go
package main

import (
	"bufio"
	"os"
)

var MAX_BIDDING = 7

func computeBidding(key_or_lock []string) []int {
	bidding := make([]int, len(key_or_lock[0]))
	for i := 0; i < len(key_or_lock[0]); i++ {
		count := 0
		for _, row := range key_or_lock {
			if i >= len(row) {
				println(i)
				println(row)
			}
			if row[i] == '#' {
				count++
			}
		}
		bidding[i] = count
	}
	return bidding
}

func parseInput() ([][]int, [][]int) {
	keys, locks := [][]int{}, [][]int{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		current := make([]string, MAX_BIDDING)
		for i := 0; i < MAX_BIDDING; i++ {
			current[i] = scanner.Text()
			scanner.Scan()
		}
		isLock := true
		for _, chr := range current[0] {
			if chr != '#' {
				isLock = false
			}
		}
		bidding := computeBidding(current)
		if isLock {
			locks = append(locks, bidding)
		} else {
			keys = append(keys, bidding)
		}
	}

	return keys, locks
}

func fits(key []int, lock []int) bool {
	for i := 0; i < len(key); i++ {
		if key[i]+lock[i] > MAX_BIDDING {
			return false
		}
	}
	return true
}

func main() {
	keys, locks := parseInput()
	pairs := 0
	for _, key := range keys {
		for _, lock := range locks {
			if fits(key, lock) {
				pairs++
			}
		}
	}
	println(pairs)
}

File: 2024/25/25.rb
#!/usr/bin/env ruby

MAX_BIDDING = 7

def compute_bidding(key_or_lock)
  (0...key_or_lock.first.size).map do |i|
    key_or_lock.count do |row|
      row[i] == '#'
    end
  end
end

def lock_or_key(lock_or_key, locks, keys)
  if lock_or_key.first.all? '#'
    locks << compute_bidding(lock_or_key)
  else
    keys << compute_bidding(lock_or_key)
  end
end

def read_input
  keys, locks = [[], []]
  current = []
  $stdin.readlines.each do |line|
    line = line.chomp
    if line.empty?
      lock_or_key(current, locks, keys)
      current = []
    else
      current << line.chars
    end
  end
  lock_or_key(current, locks, keys)
  [keys, locks]
end

keys, locks = read_input
pairs = keys.sum do |key|
  locks.count do |lock|
    key.each_with_index.all? do |bidding, pin|
      bidding + lock[pin] <= MAX_BIDDING
    end
  end
end
puts pairs
